clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -13603,6 +13603,19 @@ def note_unsafe_count_attributed_pointer_argument : Note<
 def warn_unsafe_single_pointer_argument : Warning<
   "unsafe assignment to function parameter of __single pointer type">,
   InGroup<UnsafeBufferUsage>, DefaultIgnore;
+def warn_unsafe_assign_to_null_terminated_pointer : Warning<
+  "unsafe %select{assignment to a pointer of|"
+                  "assignment to a parameter of|"
+                  "return of|"
+                  "conversion to|"
+                  "initialization of a pointer of|"
+                  "assignment to a pointer of|"
+                  "casting to}2 '__null_terminated' type; "
+  "only '__null_terminated' pointers, string literals, and 'std::string::c_str' calls are compatible with '__null_terminated' pointers">,
+  InGroup<UnsafeBufferUsage>, DefaultIgnore;
+def warn_unsafe_named_cast_to_null_terminated_pointer : Warning<
+  "C++ named cast to '__null_terminated' type is unsafe">,
```

Comment:
This may be overly specific: any cast to `__null_terminated` is unsafe. If I got this error message I'd try doing a C-style cast instead, which would not help.

Reply:
Wait, looking at the test cases it seems like C-style casts are deliberately allowed. What's the rationale?

Reply:
My original idea is that it should be rare for C++ programmers to cast null_terminated pointers with named casts.  And, Bounds Safety's type checking for null_terminated can only be re-used for C-Style casting. So why not force C++ programmers to just use C-Style casts for null_terminated pointers?

But now with your questioning,  I changed my mind.  I think it is too strict.  Moreover, it will be difficult to relax the restriction once this warning is added.  On the other hand, we are allowed to have some false negatives for some rare use cases and make some warnings for them later.   So I think I will remove this warning.  Thanks for bringing this up!

Reply:
I disagree that it's easier to add warnings later than relaxing later, since relaxing the rules doesn't disrupt existing adopters while new warnings do. But that said, if there's no reason for preferring one cast over the other it seems reasonable to remove it.

Reply:
Where's the test case to allow C-Style casts?